File: src/api/sessions.ts
import { invoke } from "./client";
import type { AppKind } from "./shared";

/** Read-only local session metadata. The backend never exposes source paths. */
export interface SessionMeta {
  app: AppKind;
  sessionId: string;
  title: string;
  summary: string;
  projectDir: string | null;
  createdAt: string | null;
  lastActiveAt: string | null;
  resumeCommand: string;
  alias: string | null;
  pinned: boolean;
  tags: string[];
}

export interface SessionMessage {
  id: string;
  role: string;
  content: string;
  at: string | null;
}

export interface SessionIssue {
  app: AppKind;
  message: string;
}

export interface SessionSearchHit {
  session: SessionMeta;
  messageId: string | null;
  excerpt: string;
}

export interface SessionSearchPage {
  results: SessionSearchHit[];
  total: number;
  issues: SessionIssue[];
  projects: SessionProject[];
  tags: string[];
}

export interface SessionProject {
  app: AppKind;
  projectDir: string | null;
  /** Unique sessions in this project, scoped only by the selected client. */
  count: number;
}

export interface SessionSearchRequest {
  query: string;
  app: AppKind | null;
  offset: number;
  project: Pick<SessionProject, "app" | "projectDir"> | null;
  tag: string | null;
  pinnedOnly: boolean;
  /** Inclusive RFC3339 lower bound for lastActiveAt. */
  activeAfter: string | null;
  /** Exclusive RFC3339 upper bound for lastActiveAt. */
  activeBefore: string | null;
}

export type SessionOrganizationChange =
  | { kind: "pin"; pinned: boolean }
  | { kind: "alias"; alias: string | null }
  | { kind: "setTags" | "addTags" | "removeTags"; tags: string[] };

/** Result of starting a supported CLI's resume command in a new terminal. */
export interface SessionResume {
  command: string;
  usedProjectDir: boolean;
}

export function searchSessions(request: SessionSearchRequest): Promise<SessionSearchPage> {
  return invoke("search_sessions", { request });
}

export function updateSessionOrganization(
  requests: SessionDeleteRequest[], change: SessionOrganizationChange,
): Promise<SessionMeta[]> {
  return invoke("update_session_organization", { requests, change });
}

export interface SessionBookmark {
  id: string;
  app: AppKind;
  sessionId: string;
  sessionTitle: string;
  projectDir: string | null;
  messageId: string;
  role: string;
  content: string;
  at: string | null;
  savedAt: string;
}

export function exportSessionMarkdown(app: AppKind, sessionId: string): Promise<string | null> {
  return invoke("export_session_markdown", { app, sessionId });
}

export function listSessionBookmarks(): Promise<SessionBookmark[]> {
  return invoke("list_session_bookmarks");
}

export function saveSessionBookmark(app: AppKind, sessionId: string, messageId: string): Promise<SessionBookmark> {
  return invoke("save_session_bookmark", { app, sessionId, messageId });
}

export function deleteSessionBookmark(id: string): Promise<void> {
  return invoke("delete_session_bookmark", { id });
}

export function getSessionMetadata(app: AppKind, sessionId: string): Promise<SessionMeta> {
  return invoke("get_session_metadata", { app, sessionId });
}

export function getSessionMessages(app: AppKind, sessionId: string): Promise<SessionMessage[]> {
  return invoke<SessionMessage[]>("get_session_messages", { app, sessionId });
}

export function resumeSession(app: AppKind, sessionId: string): Promise<SessionResume> {
  return invoke<SessionResume>("resume_session", { app, sessionId });
}

/** Permanently removes the local session record the backend resolves itself. */
export function deleteSession(app: AppKind, sessionId: string): Promise<void> {
  return invoke<void>("delete_session", { app, sessionId });
}

export interface SessionDeleteRequest {
  app: AppKind;
  sessionId: string;
}

/** One line of a batch deletion: removed, or the reason the record stayed. */
export interface SessionDeleteOutcome {
  app: AppKind;
  sessionId: string;
  deleted: boolean;
  error: string | null;
}

/** Deletes several records against one backend scan; each item reports its own outcome. */
export function deleteSessions(requests: SessionDeleteRequest[]): Promise<SessionDeleteOutcome[]> {
  return invoke<SessionDeleteOutcome[]>("delete_sessions", { requests });
}

File: src-tauri/src/commands/sessions.rs
use super::error::{blocking, observe, state, CommandError};
use crate::runtime_log::RuntimeLogAction;
use crate::session_manager;
use asb_core::contracts::AppKind;
use tauri::AppHandle;
use tauri_plugin_dialog::DialogExt;

/// Searches only backend-resolved session roots and returns a bounded page.
#[tauri::command]
pub async fn search_sessions(
    app_handle: AppHandle,
    request: session_manager::SessionSearchRequest,
) -> Result<session_manager::SessionSearchPage, CommandError> {
    let local = state(&app_handle)?;
    blocking(move || {
        session_manager::search_sessions(local.root(), request)
            .map_err(|message| CommandError::localized(
                "session-search-failed",
                "errors.sessions.searchFailed",
                message.clone(),
                serde_json::json!({ "detail": message }),
            ))
    })
    .await
}

/// Resolves one exact session identity independently of search pagination.
#[tauri::command]
pub async fn get_session_metadata(
    app_handle: AppHandle,
    app: AppKind,
    session_id: String,
) -> Result<session_manager::SessionMeta, CommandError> {
    let local = state(&app_handle)?;
    blocking(move || {
        session_manager::load_metadata(local.root(), app, &session_id)
            .map_err(|message| CommandError::localized(
                "session-unavailable",
                "errors.sessions.unavailable",
                message.clone(),
                serde_json::json!({ "detail": message }),
            ))
    })
    .await
}

/// The renderer selects a session, never a filesystem source or export path.
#[tauri::command]
pub async fn export_session_markdown(
    app_handle: AppHandle,
    app: AppKind,
    session_id: String,
) -> Result<Option<String>, CommandError> {
    let local = state(&app_handle)?;
    blocking(move || {
        let markdown = session_manager::export_markdown(local.root(), app, &session_id)
            .map_err(|message| CommandError::localized(
                "session-export-failed",
                "errors.sessions.exportFailed",
                message.clone(),
                serde_json::json!({ "detail": message }),
            ))?;
        let filename_id: String = session_id
            .chars()
            .filter(|character| character.is_ascii_alphanumeric() || matches!(*character, '-' | '_'))
            .take(64)
            .collect();
        let Some(destination) = app_handle
            .dialog()
            .file()
            .add_filter("Markdown", &["md"])
            .set_file_name(format!("session-{filename_id}.md"))
            .blocking_save_file()
        else {
            return Ok(None);
        };
        let path = destination.into_path().map_err(|error| {
            CommandError::localized(
                "session-export-failed",
                "errors.sessions.exportPathInvalid",
                format!("无法使用所选导出路径：{error}"),
                serde_json::json!({ "detail": error.to_string() }),
            )
        })?;
        std::fs::write(&path, markdown).map_err(|error| {
            CommandError::localized(
                "session-export-failed",
                "errors.sessions.exportWriteFailed",
                format!("无法写入会话 Markdown：{error}"),
                serde_json::json!({ "detail": error.to_string() }),
            )
        })?;
        Ok(Some(path.to_string_lossy().into_owned()))
    })
    .await
}

#[tauri::command]
pub async fn list_session_bookmarks(
    app_handle: AppHandle,
) -> Result<Vec<session_manager::SessionBookmark>, CommandError> {
    let local = state(&app_handle)?;
    blocking(move || {
        session_manager::list_bookmarks(local.root())
            .map_err(|message| CommandError::localized(
                "session-bookmarks-unavailable",
                "errors.sessions.bookmarksUnavailable",
                message.clone(),
                serde_json::json!({ "detail": message }),
            ))
    })
    .await
}

#[tauri::command]
pub async fn save_session_bookmark(
    app_handle: AppHandle,
    app: AppKind,
    session_id: String,
    message_id: String,
) -> Result<session_manager::SessionBookmark, CommandError> {
    let local = state(&app_handle)?;
    blocking(move || {
        session_manager::save_bookmark(local.root(), app, &session_id, &message_id)
            .map_err(|message| CommandError::localized(
                "session-bookmark-save-failed",
                "errors.sessions.bookmarkSaveFailed",
                message.clone(),
                serde_json::json!({ "detail": message }),
            ))
    })
    .await
}

#[tauri::command]
pub async fn delete_session_bookmark(
    app_handle: AppHandle,
    id: String,
) -> Result<(), CommandError> {
    let local = state(&app_handle)?;
    blocking(move || {
        session_manager::delete_bookmark(local.root(), &id)
            .map_err(|message| CommandError::localized(
                "session-bookmark-delete-failed",
                "errors.sessions.bookmarkDeleteFailed",
                message.clone(),
                serde_json::json!({ "detail": message }),
            ))
    })
    .await
}

/// Loads one transcript after resolving its source from the approved session
/// roots. No session file is created, modified, or deleted by this command.
#[tauri::command]
pub async fn get_session_messages(
    app: AppKind,
    session_id: String,
) -> Result<Vec<crate::session_manager::SessionMessage>, CommandError> {
    blocking(move || {
        crate::session_manager::load_messages(app, &session_id)
            .map_err(|message| CommandError::localized(
                "session-unavailable",
                "errors.sessions.unavailable",
                message.clone(),
                serde_json::json!({ "detail": message }),
            ))
    })
    .await
}

/// Starts the selected session in the platform terminal. The session source
/// is resolved by the backend; the renderer never submits a path or
/// executable command.
#[tauri::command]
pub async fn resume_session(
    app: AppKind,
    session_id: String,
) -> Result<crate::session_manager::SessionResume, CommandError> {
    observe(RuntimeLogAction::SessionResumed, async move {
        blocking(move || {
            crate::session_manager::resume_session(app, &session_id)
                .map_err(|message| CommandError::localized(
                    "session-resume-failed",
                    "errors.sessions.resumeFailed",
                    message.clone(),
                    serde_json::json!({ "detail": message }),
                ))
        })
        .await
    })
    .await
}

/// Permanently removes the one local session record the backend resolves from
/// the approved roots. Irreversible by contract; the confirmation sheet in
/// the renderer is the only gate.
#[tauri::command]
pub async fn delete_session(app_handle: AppHandle, app: AppKind, session_id: String) -> Result<(), CommandError> {
    let local = state(&app_handle)?;
    observe(RuntimeLogAction::SessionDeleted, async move {
        blocking(move || {
            crate::session_manager::delete_session(local.root(), app, &session_id)
                .map_err(|message| CommandError::localized(
                    "session-delete-failed",
                    "errors.sessions.deleteFailed",
                    message.clone(),
                    serde_json::json!({ "detail": message }),
                ))
        })
        .await
    })
    .await
}

/// Batch counterpart of `delete_session`: one scan resolves every requested
/// record and each item reports its own outcome, so an unresolved or
/// undeletable id never blocks the rest. Only an unreadable root set fails
/// the whole call.
#[tauri::command]
pub async fn delete_sessions(
    app_handle: AppHandle,
    requests: Vec<crate::session_manager::SessionDeleteRequest>,
) -> Result<Vec<crate::session_manager::SessionDeleteOutcome>, CommandError> {
    let local = state(&app_handle)?;
    observe(RuntimeLogAction::SessionDeleted, async move {
        blocking(move || {
            crate::session_manager::delete_sessions(local.root(), &requests)
                .map_err(|message| CommandError::localized(
                    "session-delete-failed",
                    "errors.sessions.deleteFailed",
                    message.clone(),
                    serde_json::json!({ "detail": message }),
                ))
        })
        .await
    })
    .await
}

#[tauri::command]
pub async fn update_session_organization(
    app_handle: AppHandle,
    requests: Vec<session_manager::SessionDeleteRequest>,
    change: session_manager::SessionOrganizationChange,
) -> Result<Vec<session_manager::SessionMeta>, CommandError> {
    let local = state(&app_handle)?;
    blocking(move || {
        session_manager::update_session_organization(local.root(), &requests, &change)
            .map_err(|message| CommandError::localized(
                "session-organization-failed",
                "errors.sessions.organizationFailed",
                message.clone(),
                serde_json::json!({ "detail": message }),
            ))
    }).await
}
